### PR TITLE
Referral scheme

### PR DESCRIPTION
**Pull request by halaprix**
# [Referral scheme](https://#)

  
## Changes 👷‍♀️
- add the referral scheme dashboard, landing page, referral modal
- add new db tables handle the referral system
- add new api routes to handle db data read/write

image
image
image

  
## How to test 🧪
1. Clone the redeemer repo from [here](https://github.com/OasisDEX/oasis_merkle) to get the contract source and db helper.
2. Start the Oasis db and push the new schema `npx prisma db push` in contract repo. 
3. Start the hardhat node `npx hardhat node --max-memory 4192 --fork https://<your favourite RPC link>`
4. deploy the contract to the local network `npx hardhat run --network localhost scripts/seed-contract-and-db.ts` this will also seed the database with required data 
image
5. start `borrow` dev server
6. go to `http://localhost:3000/referrals/?network=hardhat` for the landing page (or click banner on main page) - `Get Started` to go the referral dashboard
7. if you want to test referring other users you can use a ref link leading to the home page `http://localhost:3000/?ref=0x70997970C51812dc3A010C7d01b50e0d17dc79C8&network=hardhat`

### CODE REVIEW

File: blockchain/addresses/mainnet.json [hunk 1]
```diff
@@ -305,5 +305,6 @@
   "OPTIMISM_GOV_RELAY": "0x09B354CDA89203BB7B3131CC728dFa06ab09Ae2F",
   "ARBITRUM_DAI_BRIDGE": "0xD3B5b60020504bc3489D6949d545893982BA3011",
   "ARBITRUM_ESCROW": "0xA10c7CE4b876998858b1a9E12b10092229539400",
-  "ARBITRUM_GOV_RELAY": "0x9ba25c289e351779E0D481Ba37489317c34A899d"
+  "ARBITRUM_GOV_RELAY": "0x9ba25c289e351779E0D481Ba37489317c34A899d",
+  "MERKLE_REDEEMER": "0xC66AB83418C20A65C3f8e83B3d11c8C3a6097b6F"
```
Comment (sennett) [inline blockchain/addresses/mainnet.json:309]: This file is copied and pasted from https://chainlog.makerdao.com/ and not edited directly.  We should move this config to `blockchain/config.ts`.

File: blockchain/addresses/goerli.json [hunk 2]
```diff
@@ -276,5 +276,6 @@
   "RWA006_A_INPUT_CONDUIT": "0xd2Ef07535267D17d2314894f7821A43e9700A02e",
   "RWA006_A_OUTPUT_CONDUIT": "0xd2Ef07535267D17d2314894f7821A43e9700A02e",
   "PROXY_PAUSE_ACTIONS": "0x8D1187FCa9A104211bd25c689C08718AD8730C83",
-  "PROXY_DEPLOYER": "0xc9476Fd378de5b0de5C4280D4323f6F89f723c15"
+  "PROXY_DEPLOYER": "0xc9476Fd378de5b0de5C4280D4323f6F89f723c15",
+  "MERKLE_REDEEMER": "0x0A0647e629A0825353B76dEeC232b29df960ac2d"
```
Comment (sennett) [inline blockchain/addresses/goerli.json:280]: This file is copied and pasted from https://chainlog.makerdao.com/ and not edited directly.  We should move this config to `blockchain/config.ts`.

File: blockchain/addresses/mainnet.json (hunk 1 above)
Comment (halaprix) [reply blockchain/addresses/mainnet.json]: good to know :) fixed

File: blockchain/addresses/goerli.json (hunk 2 above)
Comment (halaprix) [reply blockchain/addresses/goerli.json]: fixed

File: blockchain/calls/txMeta.ts [hunk 3]
```diff
@@ -15,4 +15,5 @@ export enum TxMetaKind {
   addTrigger = 'addTrigger',
   removeTrigger = 'removeTrigger',
   claimReward = 'claimReward',
+  claim = 'claim',
```
Comment (paszkowskiDamian) [inline blockchain/calls/txMeta.ts:18]: maybe we can name it better, we already have `claimReward` and `reclaim`.

Comment (halaprix) [reply blockchain/calls/txMeta.ts:18]: makes sense, renamed to `claimReferralFees`

File: blockchain/calls/merkleRedeemer.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  proof: Array<BytesLike>
+  week: BigNumber
+  amount: BigNumber
+}
+
+export type ClaimMultipleData = {
+  kind: TxMetaKind.claimReferralFees
+  weeks: any
+  amounts: any
+  proofs: any
```
Comment (paszkowskiDamian) [inline blockchain/calls/merkleRedeemer.ts:18]: Can it be more specific?

File: features/referralOverview/referralLocal.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { Observable, of } from 'rxjs'
+import { delay } from 'rxjs/operators'
+
+export function checkReferralLocalStorage$(): Observable<string | null> {
+  let referrer = localStorage.getItem(`referral`)
+  if (referrer === 'null') {
+    referrer = null
+  }
+  return of(referrer).pipe(delay(1000))
```
Comment (paszkowskiDamian) [inline features/referralOverview/referralLocal.ts:9]: Why did you add delay?

File: helpers/useFeatureToggle.ts [hunk 6]
```diff
@@ -13,17 +13,19 @@ export type Feature =
   | 'StopLossRead'
   | 'StopLossWrite'
   | 'StopLossOpenFlow'
+  | 'Referrals'
 
 const configuredFeatures: Record<Feature, boolean> = {
   TestFeature: false, // used in unit tests
-  AnotherTestFeature: true, // used in unit tests
+  AnotherTestFeature: true, // used in unit sssssssssssssssssssssssssssstests
```
Comment (paszkowskiDamian) [inline helpers/useFeatureToggle.ts:20]: whaaaaaaat? 🤔

File: pages/api/user/claims/[address].tsx [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import { WeeklyClaim } from '@prisma/client'
+import { withSentry } from '@sentry/nextjs'
+import { NextApiRequest, NextApiResponse } from 'next'
+// import { userInfo } from 'os'
```
Comment (paszkowskiDamian) [inline pages/api/user/claims/[address].tsx:4]: Do you need this one?

**Comment by paszkowskiDamian** [comment]
`https://oasis.app/?ref=0xd0612c759fC6C512B7C3abceB2366C4f4096451C`

I don't really get how it works. I've created this link. Opened the link in new browser, connected some other wallet and I didn't get any popup.

File: features/referralOverview/referralLocal.ts (hunk 5 above)
Comment (halaprix) [reply features/referralOverview/referralLocal.ts]: This was quite a coincidental thing ( I've copied  `checkAcceptanceLocalStorage$` ) - it delayed the `referrer` being emitted and that caused the entire `userReferral` pipe to fire with a 1s delay - that lead to `useEffect` in `HomePageView` being executed when `router.query` was ready.. uff..

I've removed the `delay` and added `router.isReady` to the userEffect's dependency array instead.

File: helpers/useFeatureToggle.ts (hunk 6 above)
Comment (halaprix) [reply helpers/useFeatureToggle.ts]: Looked suspicious to me too ;p fixed

File: pages/api/user/claims/[address].tsx (hunk 7 above)
Comment (halaprix) [reply pages/api/user/claims/[address].tsx]: cleaned

**Comment by halaprix** [comment]
> 

I've added `window.origin` instead of hardcoded `https://oasis.app/` to ease the testing:
image

**Review by paszkowskiDamian — APPROVED** [review]
